internal/management/controller/slots/runner/runner.go

```diff
@@ -177,8 +177,11 @@ func synchronizeReplicationSlots(
 		}
 	}
 	for _, slot := range slotsInLocal.Items {
-		// We delete the slots not present on the primary or old HA replication slots.
-		if !slotsInPrimary.Has(slot.SlotName) || slot.SlotName == mySlotName {
+		// Delete slots on standby with wrong state:
+		//  * slots not present on the primary
+		//  * the slot used by this node
+		//  * slots holding xmin (this can happen on a former primary)
+		if !slotsInPrimary.Has(slot.SlotName) || slot.SlotName == mySlotName || slot.HoldsXmin {
```

Comment:
Want to double check this, because I'm worried that this loop could also run in the primary.
And it could by this logic delete a slot because of having `xmin`.

Reply:
OK, answering my own question.
This loop does not get executed in the primary, but the reason is subtle.
There's
```
func (r *InstanceReconciler) configureSlotReplicator(cluster *apiv1.Cluster) {
	switch r.instance.PodName {
	case cluster.Status.CurrentPrimary, cluster.Status.TargetPrimary:
		r.instance.ConfigureSlotReplicator(nil)
	default:
		r.instance.ConfigureSlotReplicator(cluster.Spec.ReplicationSlots)
	}
}
```
If we changed things, we might inadvertently run the synchronizer in the primary.
Going to see if I think of something.

Reply:
The whole slot synchronization subsystem is based on the assumption that it only runs on a non-primary and non-designated-primary instance.

During a promotion, synchronization is stopped for both the former and the new primary.

Even if you imagine a way to force it to run on the primary once, it will result in no changes because the source and destination are equal (so no updates). When a slot has the xmin set, it means it is active, so we will never drop it.

Reply:
> it will result in no changes because the source and destination are equal (so no updates)
This loop `for _, slot := range slotsInLocal.Items` does not check the slots are the same, though.
But, I am reassured that we could not delete an active slot. Thanks for the clarification.